lib/machines.js

```diff
@@ -1113,6 +1114,93 @@ function dnsify(str) {
     return (str.toLowerCase().replace(/[^a-z0-9-]+/g, '-'));
 }
 
+// blatently stolen from sdc-vmapi lib/common/validation.js
+
+function validVolumeName(volumeName) {
+    var VALID_VOLUME_NAME_REGEXP = /^[a-zA-Z0-9][a-zA-Z0-9_\.\-]+$/;
+    if (volumeName === undefined || volumeName === null ||
+        !VALID_VOLUME_NAME_REGEXP.test(volumeName)) {
+        return false;
+    }
+
+    return true;
+}
+
+function validVolumeMountpoint(mountpoint) {
+    if (typeof (mountpoint) !== 'string') {
+        return false;
+    }
+
+    /*
+     * Must start with '/'.
+     */
+    if (mountpoint.length === 0 || mountpoint[0] !== '/') {
+        return false;
+    }
+
+    /*
+     * Must not contain the null character.
+     */
+    if (mountpoint.indexOf('\0') !== -1) {
+        return false;
+    }
+
+    /*
+     * Must contain at least one character that is not '/'.
+     */
+    if (mountpoint.search(/[^\/]/) === -1) {
+        return false;
+    }
+
+    return true;
+}
+
+// stops at first error
+function validateVolume(volume) {
+    var VALID_VOLUME_MODES = ['ro', 'rw'];
+
+    if (volume.type !== undefined && volume.type !== 'tritonnfs') {
+        return (new InvalidArgumentError('unknown volume type "' + volume.type
+            + '"'));
+    }
+
+    if (volume.mode !== undefined &&
+        VALID_VOLUME_MODES.indexOf(volume.mode) === -1) {
+
+        return (new InvalidArgumentError('volume mode for volume '
+            + util.inspect(volume) + ' must be one of '
+            + VALID_VOLUME_MODES.join(', ')));
+    }
+
+    if (!validVolumeName(volume.name)) {
+        return (new InvalidArgumentError('invalid volume name for volume '
+            + util.inspect(volume)));
+    }
+
+    if (!validVolumeMountpoint(volume.mountpoint)) {
+        return (new InvalidArgumentError('invalid mount point for volume '
+            + util.inspect(volume)));
+    }
+
+    return undefined;
+}
+
+function validateVolumes(volumes) {
+    var idx = 0;
+    var validationErr;
+
+    assert.arrayOfObject(volumes, 'volumes');
+
+    for (idx = 0; idx < volumes.length; idx++) {
+        validationErr = validateVolume(volumes[idx]);
+        if (validationErr) {
+            break;
```

Comment:
Same here: we early out on the first validation error, should we instead validate all input and report all relevant errors?

Reply:
Yep.